# Функциональное программирование

Функции в Python такие же объекты как, например, строки, списки или классы. Функции можно передавать в другие функции, их можно возвращать из функции.

In [2]:
def caller(func, params):
    return func(*params) # * - распаковка элементов списка в последовательность аргументов

def printer(name, origin):
    print('I\'m {} of {}!'.format(name, origin))

# Функция передается в другую функцию. Функция это объект
caller(printer, ['Moana', 'Motunui'])

I'm Moana of Motunui!


## Замыкания (closure)

In [3]:
def get_multiplier():
    def inner(a, b):
        return a * b
    return inner # Возвращаем внутреннюю функцию inner

multiplier = get_multiplier() # Получаем какую-то возвращаемую функцию из функции get_multiplier()
print(multiplier.__name__)    # Имя функции стало inner, вместо get_multiplier
print(multiplier(10, 11))     # 110

inner
110


Ситуация с переменной `number` и возвратом функции называется - замыканием:

In [4]:
def get_multiplier(number):
    def inner(a):
        return a * number
    return inner # Возвращаем внутреннюю функцию inner замкнутую на переменную number

multiplier = get_multiplier(2)
print(multiplier(10)) # 20

20


## Функция map()

Функция `map` - для каждого элемента в списке выполнит опеределенную функцию:

In [5]:
def squarify(a):
    return a ** 2 # Возведение в квадрат

#           map(func, [])
print( list(map(squarify, range(5))) )

[0, 1, 4, 9, 16]


Что делает функция `map()`, можно посмотреть на этом коде:

In [7]:
squared_list = []

for i in range(5):
    squared_list.append(squarify(i))

print(squared_list) # Аналогичный результат

[0, 1, 4, 9, 16]


Задача: Функция превращает список чисел в список строк:

In [5]:
list(map(str, range(5)))

['0', '1', '2', '3', '4']

## Функция filter()

Функция `filter()` позволяет фильтровать по какому-то предикату итерабельный объект:

In [10]:
def is_positive(a):
    return a > 0

print( list(filter(is_positive, range(-5, 5))) )

[1, 2, 3, 4]


Что делает функция `filter()`, можно посмотреть на этом коде:

In [11]:
positive_list = []

for i in range(-5, 5):
    if is_positive(i):
        positive_list.append(i)

print(positive_list) # Аналогичный результат

[1, 2, 3, 4]


## lambda (Анонимные функции)

Иногда, функция нужна лишь для одноразового применения и ее декларация в коде будет излишним. Для этого подходит анонимная функция `lambda`:

| ключевое слово | входящие аргументы | двоеточие | return |
|----------------|--------------------|-----------|--------|
|     lambda     |          x         |      :    | x ** 2 |


In [2]:
print(type(lambda x: x ** 2))

<class 'function'>


Примеры использования:

In [3]:
list(map(lambda x: x ** 2, range(5)))

[0, 1, 4, 9, 16]

In [4]:
list(filter(lambda x: x > 0, range(-2, 3)))

[1, 2]

## Модуль functools

### Функция reduce

In [10]:
from functools import reduce

def multiply(a, b):
    return a * b

print( reduce(multiply, [1, 2, 3, 4, 5]) )

120


Что происходит? Вначале, в функцию `multiply` попадают два первых значения `[1, 2]`, они перемножаются и результат запоминается. Далее, этот результат и следующее число в списке вновь передаются в функцию `multiply`:

[1, 2] = 2  
[**2**, 3] = 6  
[**6**, 4] = 24  
[**24**, 5] = 120

Аналогичный пример, но с использованием не внешней, а анонимной функции:

In [8]:
reduce(lambda x,y: x * y, range(1, 6))

120

### Функция partial

Позволяет частично, заранее, перед вызовом целевой функции, определять ее аргументы.  
Например, это может быть полезно для логера, когда в функции определяется конкретный какой-то файл.

In [11]:
from functools import partial

def greeter(person, greeting):
    return('{}, {}!'.format(greeting, person))

heir = partial(greeter, greeting='Hi')
helloer = partial(greeter, greeting='Hello')

print(heir('brother')) # Hi, brother!
print(helloer('sir'))  # Hello, sir!

Hi, brother!
Hello, sir!


## Списочные выражения. Генератор

In [12]:
squared_list = []

for i in range(10):
    squared_list.append(i ** 2)
    
print(squared_list)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


Запись выше многословна и в Python это не приветствуется, для этого существует генератор списка:

In [13]:
squared_list = [i ** 2 for i in range(10)]

print(squared_list)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


В списочном выражении можно применить условие, например, заполнить список только четными числами:

In [14]:
squared_list = [i for i in range(10) if i % 2 == 0]

print(squared_list)

[0, 2, 4, 6, 8]


Точно также мы можем определять словари:

In [15]:
square_map = {i: i ** 2 for i in range(5)}

print(square_map)

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16}


или множества:

In [16]:
reminders_set = {i % 10 for i in range(100)}

print(reminders_set)

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}


Запись определенная в списке, словаре или множестве возвращает генератор, а не список:

In [19]:
print( type(i for i in range(5)) )

<class 'generator'>


## Функция zip

Функция zip позволяет склеить 2 итерабельных объекта:

In [29]:
num_list = range(5)                        # [0, 1, 2, 3, 4]
squared_list = [x ** 2 for x in num_list]  # [0, 1, 4, 9, 16]

print( list(zip(num_list, squared_list)) )

[(0, 0), (1, 1), (2, 4), (3, 9), (4, 16)]


## Итоги. Функциональное программирование:
* Функции - объекты первого класса
* map, filter, reduce, partial
* lambda - анонимные функции
* Списочные выражения